spacy/tokens/doc.pyx

```diff
@@ -1924,6 +2042,180 @@ cdef int [:,:] _get_lca_matrix(Doc doc, int start, int end):
     return lca_matrix
 
 
+@cython.boundscheck(False)  # Deactivate bounds checking
+cdef void _set_prefix_lengths(
+    const unsigned char* tok_str,
+    const int tok_str_l,
+    const int p_max_l, 
+    unsigned char* pref_l_buf,
+) nogil:
+    """ Populate *pref_l_buf*, which has length *pref_l*, with the byte lengths of the first *pref_l* characters within *tok_str*. 
+        Lengths that are greater than the character length of the whole word are populated with the byte length of the whole word.
+
+        tok_str: a UTF-8 representation of a string.
+        tok_str_l: the length of *tok_str*.
+        p_max_l: the number of characters to process at the beginning of the word.
+        pref_l_buf: a buffer of length *p_max_l* in which to store the lengths. The calling code ensures that lengths
+            greater than 255 cannot occur.
+    """
+    cdef int tok_str_idx = 1, pref_l_buf_idx = 0
+
+    while pref_l_buf_idx < p_max_l:
+        if (tok_str_idx >= tok_str_l 
+            or 
+            ((tok_str[tok_str_idx] & 0xc0) != 0x80)  # not a continuation character
+        ):
+            pref_l_buf[pref_l_buf_idx] = tok_str_idx
+            pref_l_buf_idx += 1
```

Comment:
If you'd end up using pointers here, you don't need the `pref_l_buf_idx` variable, you could increment `pref_l_buf`, so that the pointer points to the next element.